lib/web_catalog/api_extractor.es6.js

```diff
@@ -140,511 +242,428 @@ foam.CLASS({
+            if (!names.includes(name)) names.push(name);
+            this.libs_[id] = this.apisConcat_((this.libs_[id] || []), names);
           }
-        } else if (!og.isFunction(id)) {
-          // Object is not a first level object or prototype function,
-          // nor is it a function.
-          let ownProperty = this.getObjectProperties_(og, id);
-          if (this.setMinus_(ownProperty,
-                             this.builtInObjectProperties).length > 0) {
-            let className = this.getClassName_(id, og);
-            if (((!className) || className === 'Object') && firstLevel) {
-
-              // If object is at first level and its class name is "Object",
-              // this object is considered as an object library,
-              // such as MATH, console.
-              let prototypePropertyId = og.lookup('prototype', id);
-              // In Firefox, Edge, and Safari (at least),
-              // objects can have prototype properties.
-              let prototypePropertyProps = [];
-              if (prototypePropertyId) {
-                // Capture all APIs descending from <current object>.prototype
-                // that do not belong to a different interface.
-                //
-                // Walk the prototype chain of <current object>.prototype until:
-                // (1) the prototype's class name changes (i.e., prototype
-                //     belongs to another API; OR
-                // (2) Object.prototype is reached (this is necessary because
-                //     the current class name may be "Object").
-                const objectPrototypePropertyId = og.lookup('Object.prototype');
-                let prototypePropertyClassName =
-                    this.getClassName_(prototypePropertyId, og);
-                while (!og.isType(prototypePropertyId) &&
-                       prototypePropertyId !== objectPrototypePropertyId &&
-                       prototypePropertyClassName === className) {
-                  prototypePropertyProps = prototypePropertyProps.concat(
-                      this.getObjectProperties_(og, prototypePropertyId));
-                  prototypePropertyId = og.getPrototype(prototypePropertyId);
-                  prototypePropertyClassName =
-                      this.getClassName_(prototypePropertyId, og);
+        }
+
+        // Gather APIs from libraries function/prototype pairs.
+        for (const id of allOgIds) {
+          const isCtor = this.fns_.hasOwnProperty(id);
+          const isLib = this.libs_.hasOwnProperty(id);
+          const isCtorProto = this.protos_.hasOwnProperty(id);
+
+          foam.assert(!((isCtor && isLib) || (isCtor && isCtorProto) ||
+                        (isLib && isCtorProto)),
+                      'ApiExtractor: Object with multiple personalities');
+
+          // Functions are interfaces, not prototypes. Just store properties and
+          // carry on.
+          if (isCtor) {
```

Comment:
Since this will only ever be true for elements of this.fns_ (and the same for the other branches), why not split into multiple loops?